### ISSUE TITLE
KeysMatcher for matching maps

### ISSUE THREAD

**Issue by xanderflood** [onsi/gomega#301]
I'd like to have functionality essentially identical to `MatchFields`, but for maps instead of structs. I have something written to this effect based closely on `MatchFields` already, but I'm not entirely sure that `gstruct` doesn't provide a way to do this already.

Any thoughts on workarounds/how to address this with the existing interface?

Any concerns on how such a solution should actually differ from `MatchFields`, or would it be reasonable to mimic that behavior as closely as possible?

**Comment by blgm** [comment]
As far as I can tell, gstruct seems quite tied to structs, and does not work for maps.

I like the idea of similar functionality for testing maps.  Ideally gstruct would just work on maps, but that might not be realistic.

You mentioned that you already had an implementation.  Is that publicly available?  Are there any examples of it in use?  I find that centring a discussion around some existing code helps to keep it focused.

**Comment by blgm** [comment]
The standard Gomega matchers allow things like:
```go
m := map[string]int{"Foo": 3, "BazFoo": 4}
Expect(m).To(HaveKeyWithValue(
  MatchRegexp(`.+Foo$`),
  BeNumerically(">", 3)
))
```
But this works on one key at a time.

**Comment by xanderflood** [comment]
https://github.com/xanderflood/gomega/tree/xf/match-keys implementation is here. The logic is pretty much a clone of MatchFields, and so are the tests. I can provide some examples of it in use at a later time

@blgm That works generally and is not much different from how I've done it in the past. The main caveat I ran into that led me to open the issue is when I want to apply matchers to the keys of a map while nested inside of a larger `gstruct` expression, in which case I can't list them out one at a time

**Comment by williammartin** [comment]
@blgm Did you have any opinion on accepting a change for `MatchKeys`?

**Comment by xanderflood** [comment]
Opened https://github.com/onsi/gomega/pull/317 just so that it's formally here

**Comment by williammartin** [comment]
Merged #317
